knitr/planetary_motion/planetary_motion.rmd

```diff
@@ -0,0 +1,735 @@
+The hope is that a fix in this simple context translates into a fix in a more sophisticated setting.
+This turns out to be quite true here where a simple one-parameter model allows us to understand the fundamentally multimodal nature of the model.
+
+There are many ways to simplify a model.
+A general approach is to fix some of the parameters, which we can easily do when working with simulated data.
+We fix all the parameters, except for $k$.
+We now want to characterize the posterior distribution,
+$$
+   p(k \mid q_\mathrm{obs}),
+$$
+by fitting the model with Stan.
+Since $k$ is a scalar, we could use quadrature but our goal is to understand the pathologies that frustrate our inference algorithm, so we stick to MCMC.
+As a prior, we use
+$$
+  k \sim \mathrm{Normal}^+(0, 1),
+$$
+which is fairly uninformative.
+
+We fit 8 chains in parallel.
+```{r }
+chains <- 8
+```
+
+```
+fit <- mod$sample(data = list(n = n, q_obs = q_obs),
+                  chains = chains, parallel_chains = chains,
+                  iter_warmup = 500,
+                  iter_sampling = 500,
+                  seed = 123, save_warmup = TRUE)
+```
+
+```{r }
+# The model takes a while to run, so we read in the saved output.
+r_fit1 <- readRDS("saved_fit/fit1.RDS")
+get_elapsed_time(r_fit1)
+```
+
+The first notable pathology is that some of the chains take a much longer time to run. The difference is not subtle...
+
+Let's examine the summary.
+```{r }
+summary(r_fit1, pars = c("lp__", "k"), probs = c())[1]
+```
+$\hat R \gg 1$. 
+Wow, these numbers are dramatic!
+Clearly the chains are not mixing and we can visualize this using trace plots.
+```{r fig.height=3}
+traceplot(r_fit1, pars = c("lp__", "k"))
+```
+
+
+Notice that chains 3 and 4, which ran for longer than 1000 seconds, sample around the largest values of $k$ and also produce the lowest log posterior distribution.
+It's not uncommon for issues to come in bulks.
+By contrast, chains that sample around $k = 1$ run in ~2 seconds, and produce a high log posterior.
+
+At this point, we may formulate several hypothesis as to what may be happening:
+
+* The model is degenerate and cannot resolve the correct value of $k$.
+* The chains get stuck at local modes and cannot escape, even if the log posterior is much higher near $k = 1$.
+* The chains get stuck on flat surfaces, and the gradient doesn't guide the chain back to $k \approx 1$.
+* The numerical integrator struggles to accurately solve the ODEs for large values of $k$. This means the ODE integrator is slower or even inaccurate. 
+In the latter case, this also means our gradient calculations, and computation of HMC could be wrong.
+
+## Is the model degenerate?
+
+Bearing a slight abuse of language, we use "degenerate" to mean that various values of $k$ roughly produce the same data generating process.
+We can check for degeneracy by looking at the  _posterior predictive checks_, split across chains.
```

Comment:
When you see multiple modes, you wonder if it's because the model is not identifiable or "degenerate". The point of this section is to disprove this idea.